Add-ons (#5950)
scrapy/scrapy: docs/topics/addons.rst

.. method:: check_configuration(config, crawler)

    This method is called when the :class:`~scrapy.crawler.Crawler` has been
    fully initialized, immediately before it starts crawling. You can perform
    additional dependency and configuration checks here.

    :param config: Configuration of this add-on
    :type config: ``dict``

    :param crawler: Fully initialized Scrapy crawler
    :type crawler: :class:`~scrapy.crawler.Crawler`

Add-on base class
=================

Scrapy comes with a built-in base class for add-ons which provides some
convenience functionality: the add-on configuration can be exposed into
Scrapy's settings via :meth:`~scrapy.addons.Addon.export_config`, configurable
via :attr:`~scrapy.addons.Addon.default_config` and
:attr:`~scrapy.addons.Addon.config_mapping`.

By default, the base add-on class will expose the add-on configuration into
Scrapy's settings namespace, in upper case. It is
easy to write your own functionality while still being able to use the
convenience functions by overwriting
:meth:`~scrapy.addons.Addon.update_settings`.

.. module:: scrapy.addons
   :noindex:

.. autoclass:: Addon

>    :members:

Reviewer: I am not strongly against it, but we could probably do without a base class at all.

Author: If we remove the per-addon automatic setting names then it's no longer useful, yeah.

Author: > assuming addons are Scrapy components, meaning they can define a from_crawler class method

They are currently not Scrapy components though, we need to add a `create_instance` call to AddonManager for this to work.

Add-on examples
===============

Set some basic configuration using the :class:`Addon` base class::

    from scrapy.addons import Addon

    class MyAddon(Addon):
        name = 'myaddon'

        def update_settings(self, config, settings):

>             super().update_settings(settings)

Reviewer 2: the signature looks different

            settings["ITEM_PIPELINES"]["path.to.mypipeline"] = 200
            settings["DNSCACHE_ENABLED"] = True

Check dependencies::

    from scrapy.addons import Addon

    class MyAddon(Addon):
        name = 'myaddon'

        def update_settings(self, config, settings):
            try:
                import boto
            except ImportError:
                raise RuntimeError("myaddon requires the boto library")
            super().update_settings(settings)

Check configuration of fully initialized crawler (see
:ref:`topics-api-crawler`)::

    class MyAddon(object):
        name = 'myaddon'

        def update_settings(self, config, settings):
            super().update_settings(settings)
            settings.set('DNSCACHE_ENABLED', False, priority='addon')

        def check_configuration(self, config, crawler):
            if crawler.settings.getbool('DNSCACHE_ENABLED'):
                # The spider, some other add-on, or the user messed with the
                # DNS cache setting

>                 raise ValueError("myaddon is incompatible with DNS cache")

Reviewer: I am not sure whether or not we should encourage this approach of post-setting-update checking on add-ons. But this is one example where I think we don’t need a method. For example (untested):

```python
from logging import getLogger

from scrapy.signals import spider_opened

logger = getLogger(__name__)

class MinimalisticApproachAddon:
    
    @classmethod
    def from_crawler(cls, crawler):
        return cls(crawler)        

    def __init__(self, crawler):
        crawler.signals.connect(self.spider_opened, signal=spider_opened)

    def update_settings(self, settings):
        super().update_settings(settings)
        settings.set('DNSCACHE_ENABLED', False, priority='addon')
        
    def spider_opened(self, spider):
        if spider.settings.getbool('DNSCACHE_ENABLED'):
            logger.error(f"DNSCACHE_ENABLED is enabled, and {self.__class__} does not work with it.")
            return spider.crawler.engine.close_spider()
```

.. _topics-addons:

=======
Add-ons
=======

Scrapy's add-on system is a framework which unifies managing and configuring
components that extend Scrapy's core functionality, such as middlewares,
extensions, or pipelines. It provides users with a plug-and-play experience in
Scrapy extension management, and grants extensive configuration control to
developers.

Activating and configuring add-ons
==================================

Add-ons and their configuration live in Scrapy's
:class:`~scrapy.addons.AddonManager`. During a :class:`~scrapy.crawler.Crawler`
initialization the add-on manager will read a list of enabled add-ons from your
``ADDONS`` setting.

The ``ADDONS`` setting is a dict in which every key is an addon class or its
import path and the value is its priority.

This is an example where two add-ons are enabled in a project's
``settings.py``::

    ADDONS = {
        'path.to.someaddon': 0,
        path.to.someaddon2: 1,
    }

Writing your own add-ons
========================

> Add-ons are (any) Python objects that include the following method:

Reviewer: Even if this may be technically true, shouldn’t we encourage them being classes from the docs?

.. method:: update_settings(settings)

    This method is called during the initialization of the
    :class:`~scrapy.crawler.Crawler`. Here, you should perform dependency checks
    (e.g. for external Python libraries) and update the
    :class:`~scrapy.settings.Settings` object as wished, e.g. enable components
    for this add-on or set required configuration of other extensions.

    :param settings: The settings object storing Scrapy/component configuration
    :type settings: :class:`~scrapy.settings.Settings`

They can also have the following method:

.. classmethod:: from_crawler(cls, crawler)
   :noindex:

   If present, this class method is called to create an addon instance
   from a :class:`~scrapy.crawler.Crawler`. It must return a new instance

>    of the addon. Crawler object provides access to all Scrapy core

Reviewer: ```suggestion
   of the addon. The crawler object provides access to all Scrapy core
```

>    components like settings and signals; it is a way for pipeline to
>    access them and hook its functionality into Scrapy.

Reviewer: Is this what was meant?

```suggestion
   components like settings and signals; it is a way for the addon to
   access them and hook its functionality into Scrapy.
```

   :param crawler: The crawler that uses this addon
   :type crawler: :class:`~scrapy.crawler.Crawler`

The settings set by the addon should use the ``addon`` priority (see
:ref:`populating-settings` and :func:`scrapy.settings.BaseSettings.set`)::

    class MyAddon:
        def update_settings(self, settings):
            settings.set("DNSCACHE_ENABLED", True, "addon")

This allows users to override these settings in the project or spider
configuration. This is not possible with settings that are mutable objects,
such as the dict that is a value of :setting:`ITEM_PIPELINES`. In these cases
you can provide an addon-specific setting that governs whether the addon will
modify :setting:`ITEM_PIPELINES`::

    class MyAddon:
        def update_settings(self, settings):
            if settings.getbool("MYADDON_ENABLE_PIPELINE"):
                settings["ITEM_PIPELINES"]["path.to.mypipeline"] = 200

Fallbacks
---------

> Some components provided by addons need to fallback to "default"

Reviewer: ```suggestion
Some components provided by addons need to fall back to "default"
```

it doesn't handle via the default download handler, or a stats collector that
includes some additional processing but otherwise uses the default stats
collector. And it's possible that a project needs to use several custom
components of the same type, e.g. two custom download handlers that support
different kinds of custom requests and still need to use the default download
handler for other requests. To make such use cases easier to configure, we
recommend that such custom components should be written in the following way:

1. The custom component (e.g. ``MyDownloadHandler``) shouldn't inherit from the
   default Scrapy one (e.g.
   ``scrapy.core.downloader.handlers.http.HTTPDownloadHandler``), but instead
   be able to load the class of the fallback component from a special setting
   (e.g. ``MY_FALLBACK_DOWNLOAD_HANDLER``), create an instance of it and use
   it.
2. The addons that include these components should read the current value of
   the default setting (e.g. ``DOWNLOAD_HANDLERS``) in their
   ``update_settings()`` methods, save that value into the fallback setting
   (``MY_FALLBACK_DOWNLOAD_HANDLER`` mentioned earlier) and set the default

>    setting to the component provided byt the addon (e.g.

Reviewer: ```suggestion
   setting to the component provided by the addon (e.g.
```

        def from_crawler(cls, crawler):
            return cls(crawler)

        def update_settings(self, settings):
            ...

Use a fallback component:

.. code-block:: python

    from scrapy.core.downloader.handlers.http import HTTPDownloadHandler

    FALLBACK_SETTING = "MY_FALLBACK_DOWNLOAD_HANDLER"

    class MyHandler:
        lazy = False

        def __init__(self, settings, crawler):
            dhcls = load_object(settings.get(FALLBACK_SETTING))
            self._fallback_handler = create_instance(
                dhcls,
                settings=None,
                crawler=crawler,
            )

        def download_request(self, request, spider):
            if request.meta.get("my_params"):
                # handle the request
                ...
            else:
                return self._fallback_handler.download_request(request, spider)

    class MyAddon:
        def update_settings(self, settings):
            if not settings.get(FALLBACK_SETTING):
                settings.set(
                    FALLBACK_SETTING,
                    settings.getwithbase("DOWNLOAD_HANDLERS")["https"],
                    "addon",
                )

>             settings["DOWNLOAD_HANDLERS"]["https"] = "path.to.MyHandler"

Reviewer 2: can you use MyHandler here directly? Or would you consider it an anti-pattern? Not sure myself :)

.. _topics-addons:

=======
Add-ons
=======

Scrapy's add-on system is a framework which unifies managing and configuring
components that extend Scrapy's core functionality, such as middlewares,
extensions, or pipelines. It provides users with a plug-and-play experience in
Scrapy extension management, and grants extensive configuration control to
developers.

Activating and configuring add-ons
==================================

> Add-ons and their configuration live in Scrapy's
> :class:`~scrapy.addons.AddonManager`. During a :class:`~scrapy.crawler.Crawler`
> initialization the add-on manager will read a list of enabled add-ons from your
> ``ADDONS`` setting.

Reviewer 2: I wonder if we should remove AddonManager here. We don't put emphasis on e.g. MiddlewareManager, as users are not goint to use this directly. 

```suggestion
During a :class:`~scrapy.crawler.Crawler`
initialization it reads a list of enabled add-ons from your
``ADDONS`` setting.
```

Reviewer: I agree with removing the mention, although I would suggest a slightly different wording:

```suggestion
During :class:`~scrapy.crawler.Crawler` initialization, the list of enabled 
add-ons is read from your ``ADDONS`` setting.
```

The ``ADDONS`` setting is a dict in which every key is an add-on class or its
import path and the value is its priority.

This is an example where two add-ons are enabled in a project's
``settings.py``::

    ADDONS = {
        'path.to.someaddon': 0,

>         path.to.someaddon2: 1,

Reviewer 2: ```suggestion
        SomeAddonClass: 1,
```

   it.
2. The add-ons that include these components should read the current value of
   the default setting (e.g. ``DOWNLOAD_HANDLERS``) in their
   ``update_settings()`` methods, save that value into the fallback setting
   (``MY_FALLBACK_DOWNLOAD_HANDLER`` mentioned earlier) and set the default
   setting to the component provided by the add-on (e.g.
   ``MyDownloadHandler``). If the fallback setting is already set by the user,
   they shouldn't change it.
3. This way, if there are several add-ons that want to modify the same setting,
   all of them will fallback to the component from the previous one and then to
   the Scrapy default. The order of that depends on the priority order in the
   ``ADDONS`` setting.

Add-on examples
===============

Set some basic configuration:

.. code-block:: python

    class MyAddon:
        def update_settings(self, settings):
            settings["ITEM_PIPELINES"]["path.to.mypipeline"] = 200
            settings.set("DNSCACHE_ENABLED", True, "addon")

Check dependencies:

.. code-block:: python

    class MyAddon:
        def update_settings(self, settings):
            try:
                import boto
            except ImportError:

>                 raise RuntimeError("MyAddon requires the boto library")

Reviewer 2: What would happen if NotConfigured error is raised? I think this is the error which components raise when a dependency is not available. Is there a reason to prefer RuntimeError in addons?

Reviewer: I have no strong opinion on it, but I think it makes sense that, if you have enabled a plugin that requires a dependency, and that dependency is missing, your crawler should fail to initialize, forcing you to either disable the plugin or install the missing dependency. Same for missing required settings.

I see `NotConfigured` as something a Scrapy component raises when it is configured but disabled through settings.

If a plugin has an `*_ENABLED` setting, then I think it would make sense to support that plugin raising `NotConfigured`, and that causing the plugin to not be used at run time, and a log message indicating so during initialization, but things working otherwise.

So I think it may make sense to add support for `NotConfigured` on the addon manager class, but I think in some cases raising something else makes sense, including this one.

Reviewer 2: So, the difference is that RuntimeError stops the spider, but NotConfigured doesn't? Does it work for addons the same way?

I'm not really familiar with these parts of the code, but we're using NotConfigured in similar cases:

* https://github.com/scrapy/scrapy/blob/7522aeed357d90ed95ee10d3b5a506f1b1049d1f/scrapy/pipelines/images.py#L61
* https://github.com/scrapy/scrapy/blob/7522aeed357d90ed95ee10d3b5a506f1b1049d1f/scrapy/extensions/feedexport.py#L176
* https://github.com/scrapy/scrapy/blob/7522aeed357d90ed95ee10d3b5a506f1b1049d1f/scrapy/core/downloader/handlers/s3.py#L20

Reviewer: I think that’s because those components are configured by default, i.e. the `_ENABLED=True` is implicit, so doing otherwise would be a bad user experience. But for opt-in components, including add-ons, I think stopping the spider may be best.

Author: OK, so nothing handles `NotConfigured` raised there.  Which is expected because it's only handled in `MiddlewareManager` (which is the base for middleware, pipeline and extension managers) and `DownloadHandlers`. We should probably add handling to AddonManager to similarly silently skip add-ons that raise it?

Author: If we do that I wonder if we should merge `AddonManager.update_settings()` into `AddonManager.load_settings()` or into `AddonManager._add()` so that `AddonManager.addons` never contains add-ons that raise `NotConfigured`. Also it doesn't seem useful to have a separate function that calls add-ons' `update_settings()` as we don't have any features that would benefit from it.

            ...

Access the crawler instance:

.. code-block:: python

    class MyAddon:
        def __init__(self, crawler) -> None:
            super().__init__()
            self.crawler = crawler

        @classmethod
        def from_crawler(cls, crawler):
            return cls(crawler)

        def update_settings(self, settings):
            ...

Use a fallback component:

.. code-block:: python

    from scrapy.core.downloader.handlers.http import HTTPDownloadHandler

>     fallback_setting = "MY_FALLBACK_DOWNLOAD_HANDLER"

Reviewer 2: should we upper-case a constant name?

    class MyHandler:
        lazy = False

        def __init__(self, settings, crawler):
            dhcls = load_object(settings.get(fallback_setting))
            self._fallback_handler = create_instance(
                dhcls,
                settings=None,
                crawler=crawler,
            )

        def download_request(self, request, spider):
            if request.meta.get("my_params"):
                # handle the request
                ...
            else:
                return self._fallback_handler.download_request(request, spider)

    class MyAddon:

>         def update_settings(self, settings):

Reviewer 2: I think this addon should also enable MyHandler here.

Scrapy extension management, and grants extensive configuration control to
developers.

Activating and configuring add-ons
==================================

Add-ons and their configuration live in Scrapy's
:class:`~scrapy.addons.AddonManager`. During Scrapy's start-up process, and
only then, the add-on manager will read a list of enabled add-ons and their
configurations from your ``ADDONS`` setting.

The ``ADDONS`` setting is a dict in which every key is an addon class or its
import path and the vaoue is its priority.

The configuration of an add-on, if necessary at all, is stored as a dictionary
setting whose name is the uppercase add-on name.

This is an example where two add-ons (in this case with one requiring no
configuration) are enabled/configured in a project's ``settings.py``::

    ADDONS = {
        'path.to.someaddon': 0,
        path.to.someaddon2: 1,
    }

    SOMEADDON = {
        'some_config': True,
    }

Enabling and configuring add-ons within Python code
---------------------------------------------------

The :class:`~scrapy.addons.AddonManager` will only read from Scrapy's settings

> *at the beginning* of Scrapy's start-up process.

Reviewer 2: This is not precise; it's not about Scrapy start-up, it's about spider/crawler startup.

Afterwards, i.e. as soon as the :class:`~scrapy.addons.AddonManager` is
populated, changing the ``ADDONS`` setting or any of the add-on

> configuration dictionary settings will have no effect.

Reviewer 2: I'm not sure why we need to document this explicitly. Isn't it how most or all other settings work?

If you want to enable, disable, or configure add-ons in Python code, for example
when writing your own add-on, you will have to use the
:class:`~scrapy.addons.AddonManager`. You can access the add-on manager through
either ``crawler.addons`` or, if you are writing an add-on, through the
``addons`` argument of the :meth:`update_addons` callback. The add-on manager
provides many useful methods and attributes to facilitate interacting with the

> add-ons framework, e.g.:

Reviewer 2: I think we can drop this feature, and consider adding it if there is a real use case.

* an :meth:`~scrapy.addons.AddonManager.add` method to load add-ons,
* the :attr:`~scrapy.addons.AddonManager.enabled` list of enabled add-ons,
* :meth:`~scrapy.addons.AddonManager.enable` and
  :meth:`~scrapy.addons.AddonManager.disable` methods,
* the :attr:`~scrapy.addons.AddonManager.configs` dictionary which holds the
  configuration of all add-ons.

In this example, we ensure that the ``httpcache`` add-on is loaded, and that
its ``expiration_secs`` configuration is set to ``60``::

    # addons is an instance of AddonManager
    if 'httpcache' not in addons:
        addons.add('httpcache', {'expiration_secs': 60})
    else:
        addons.configs['httpcache']['expiration_secs'] = 60

Writing your own add-ons
========================

Add-ons are (any) Python *objects* that provide Scrapy's *add-on interface*.
The interface is enforced through ``zope.interface``. This leaves the choice of
Python object up the developer. Examples:

* for a small pipeline, the add-on interface could be implemented in the same
  class that also implements the ``open/close_spider`` and ``process_item``

>   callbacks,

Reviewer 2: In my experience this is an anti-pattern to have a single class act like different types of Scrapy components. The issue is that Scrapy would create multiple objects of this class; in this case, one instance would be used as an addon, and another as an extension. This can lead to confusion. If there is any work done during initialization ( even logging), it'd be done twice.

* for larger add-ons, or for clearer structure, the interface could be provided

>   by a stand-alone module.

Reviewer 2: I'd say by a stand-alone class.

The absolute minimum interface consists of two attributes:

> .. attribute:: name

Reviewer 2: Not sure we need this

Author: Currently this is required because of the `AddonManager._addons` dict, to be able to access, remove or disable an addon programmatically. But I think we don't want any of these features?

Author: Another place we need the name in the current code is the autodetected name for the addon settings dict (and generating that from the class name may be problematic, especially if you want to subclass it while keeping the setting name).

    string with add-on name

.. attribute:: version

    version string (PEP-404, e.g. ``'1.0.1'``)

Of course, stating just these two attributes will not get you very far. Add-ons
can provide three callback methods that are called at various stages before the
crawling process:

.. method:: update_settings(config, settings)

    This method is called during the initialization of the
    :class:`~scrapy.crawler.Crawler`. Here, you should perform dependency checks
    (e.g. for external Python libraries) and update the
    :class:`~scrapy.settings.Settings` object as wished, e.g. enable components
    for this add-on or set required configuration of other extensions.

    :param config: Configuration of this add-on
    :type config: ``dict``

    :param settings: The settings object storing Scrapy/component configuration
    :type settings: :class:`~scrapy.settings.Settings`

> .. method:: check_configuration(config, crawler)

Reviewer 2: That's actually interesting; I wonder why does it accept crawler instead of settings, i.e. why does it check crawler state, not the final settings.

Reviewer: I can see how having a crawler object available at the point in time when this method gets called could be useful, e.g. to monkeypatch some crawler component after it has been initialized.

That said, assuming addons are Scrapy components, meaning they can define a `from_crawler` class method and keep track of it (e.g. store it in  `self.crawler`), I think that and the use of signals should be able to cover most use cases there may be for having a crawler parameter in this method, or for having this method period :slightly_smiling_face:.

    This method is called when the :class:`~scrapy.crawler.Crawler` has been
    fully initialized, immediately before it starts crawling. You can perform
    additional dependency and configuration checks here.

    :param config: Configuration of this add-on
    :type config: ``dict``

    :param crawler: Fully initialized Scrapy crawler
    :type crawler: :class:`~scrapy.crawler.Crawler`

> .. method:: update_addons(config, addons)

Reviewer 2: I think we can drop it, and add if there is a real use case.

    :param config: Configuration of this add-on
    :type config: ``dict``

    :param addons: Add-on manager holding all loaded add-ons
    :type addons: :class:`~scrapy.addons.AddonManager`

Additionally, add-ons may (and should, where appropriate) provide one or more
attributes that can be used for limited automated detection of possible
dependency clashes:

.. attribute:: requires

    list of built-in or custom components needed by this add-on, as strings.

.. attribute:: modifies

    list of built-in or custom components whose functionality is affected or
    replaced by this add-on (a custom HTTP cache should list ``httpcache`` here)

.. attribute:: provides

    list of components provided by this add-on (e.g. ``mongodb`` for an
    extension that provides generic read/write access to a MongoDB database)

The entries in the :attr:`requires` and :attr:`modifies` attributes can be add-on
names or components from other add-ons' :attr:`provides` attribute. You can
specify :pep:`440`-style information about required versions. Examples::

    requires = ['httpcache']
    requires = ['otheraddon >= 2.0', 'yetanotheraddon']

> The Python object or module that is pointed to by an add-on path (e.g. given in

Reviewer 2: I think we should support only Python objects, and only allow addons as classes.

the ``ADDONS`` setting, or given to
:meth:`~scrapy.addons.AddonManager.add`) does not necessarily have to be an

> add-on. Instead, it can provide an ``_addon`` attribute. This attribute can be

Reviewer 2: I'd remove support for this `_addons` attribute.

either an add-on or another add-on path.

Add-on base class
=================

Scrapy comes with a built-in base class for add-ons which provides some
convenience functionality:

> * basic settings can be exported via :meth:`~scrapy.addons.Addon.export_basics`,

Reviewer 2: What does "exported" mean here?

Author: Basically you set a dict attribute that will be merged into Scrapy settings

  configurable via :attr:`~scrapy.addons.Addon.basic_settings`.
* a single component (e.g. an item pipeline or a downloader middleware) can be
  inserted into Scrapy's settings via
  :meth:`~scrapy.addons.Addon.export_component`, configurable via
  :attr:`~scrapy.addons.Addon.component_type`,
  :attr:`~scrapy.addons.Addon.component_key`,
  :attr:`~scrapy.addons.Addon.component`, and the ``order`` key in
  :attr:`~scrapy.addons.Addon.default_config`.
* the add-on configuration can be exposed into Scrapy's settings via
  :meth:`~scrapy.addons.Addon.export_config`, configurable via
  :attr:`~scrapy.addons.Addon.default_config`,
  :attr:`~scrapy.addons.Addon.config_mapping`, and

>   :attr:`~scrapy.addons.Addon.settings_prefix`.

Reviewer 2: I'm not sure we need this.

By default, the base add-on class will expose the add-on configuration into

> Scrapy's settings namespace, in caps and with the add-on name prepended. It is

Reviewer 2: It seems more flexible to do it without addon name prepended.

Reviewer 2: For example: you make a sublcass of an existing addon, which configures a certain extension. Suddenly the configuration is all wrong, because the extension expects old setting names.

easy to write your own functionality while still being able to use the
convenience functions by overwriting
:meth:`~scrapy.addons.Addon.update_settings`.

.. module:: scrapy.addons
   :noindex:

.. autoclass:: Addon
   :members:

Add-on examples
===============

Set some basic configuration using the :class:`Addon` base class::

    from scrapy.addons import Addon

    class MyAddon(Addon):
        name = 'myaddon'
        version = '1.0'
        component = 'path.to.mypipeline'
        component_type = 'ITEM_PIPELINES'
        component_order = 200
        basic_settings = {

>             'DNSCACHE_ENABLED': False,

Reviewer 2: the example doesn't work according to docs, as the setting is going to be exported as MY_ADDON_DNSCACHE_ENABLED

Author: The prepending is not allowed to this attribute

        }

Check dependencies::

    from scrapy.addons import Addon

    class MyAddon(Addon):
        name = 'myaddon'
        version = '1.0'

        def update_settings(self, config, settings):
            try:
                import boto
            except ImportError:

>                 raise RuntimeError("myaddon requires the boto library")

Reviewer 2: I wonder what's the benefit of checking it here, as opposed to checking it in the component's `__init__`. If it's checked in the component, we get the same error behavior when addons system is not used.

            self.export_config(config, settings)

Enable a component that lives relative to the add-on (see
:ref:`topics-api-settings`)::

    from scrapy.addons import Addon

    class MyAddon(Addon):
        name = 'myaddon'
        version = '1.0'

>         component = __name__ + '.downloadermw.coolmw'

Reviewer 2: this is weird :) Not sure why you'd like to do it. Just import the component?

        component_type = 'DOWNLOADER_MIDDLEWARES'
        component_order = 900

Instantiate components ad hoc::

    from path.to.my.pipelines import MySQLPipeline

    class MyAddon(object):
        name = 'myaddon'
        version = '1.0'

        def update_settings(self, config, settings):
            mysqlpl = MySQLPipeline(password=config['password'])
            settings.set(
                'ITEM_PIPELINES',

>                 {mysqlpl: 200},

Reviewer 2: Hm, I don't think this example works at all; ITEM_PIPELINES expects pipeline classes, not pipeline instances as keys.

Author: So, the original PR changed load_object() to add support for types and instances. We added support for types in a separate PR but didn't add support for instances. I've removed changes to load_object() from here but this part of the doc still assumes instances are supported.

                priority='addon',
            )

> Provide add-on interface along component interface::

Reviewer 2: Anti-pattern, as 2 instances would be created.

    class MyPipeline(object):
        name = 'mypipeline'
        version = '1.0'

        def process_item(self, item, spider):
            # Do some processing here
            return item

        def update_settings(self, config, settings):
            settings.set(
                'ITEM_PIPELINES',
                {self: 200},
                priority='addon',
            )

> Enable another addon (see :ref:`topics-api-addonmanager`)::

Reviewer 2: Not sure we need it.

    class MyAddon(object):
        name = 'myaddon'
        version = '1.0'

        def update_addons(self, config, addons):
            if 'httpcache' not in addons.enabled:
                addons.add('httpcache', {'expiration_secs': 60})

Check configuration of fully initialized crawler (see
:ref:`topics-api-crawler`)::

    class MyAddon(object):
        name = 'myaddon'
        version = '1.0'

        def update_settings(self, config, settings):
            settings.set('DNSCACHE_ENABLED', False, priority='addon')

        def check_configuration(self, config, crawler):
            if crawler.settings.getbool('DNSCACHE_ENABLED'):
                # The spider, some other add-on, or the user messed with the
                # DNS cache setting
                raise ValueError("myaddon is incompatible with DNS cache")

> Provide add-on interface through a module:

Reviewer 2: Not sure why it's needed.

.. code-block:: python

    name = "AddonModule"
    version = "1.0"

    class MyPipeline(object):
        ...

    class MyDownloaderMiddleware(object):
        ...

    def update_settings(config, settings):
        settings.set(
            "ITEM_PIPELINES",
            {MyPipeline(): 200},
            priority="addon",
        )
        settings.set(
            "DOWNLOADER_MIDDLEWARES",
            {MyDownloaderMiddleware(): 800},
            priority="addon",
        )

> Forward to other add-ons depending on Python version::

Reviewer 2: This can be done in other ways; also, the example uses six.PY3, and we don't support Python 2 for a long time. I'd drop this feature and the example as well.
